spring-cloud-gcp-security-firebase/src/main/java/com/google/cloud/spring/security/firebase/FirebaseJwtTokenDecoder.java

```diff
@@ -68,25 +67,20 @@
 	private final OAuth2TokenValidator<Jwt> tokenValidator;
 	private final Logger logger = LoggerFactory.getLogger(FirebaseJwtTokenDecoder.class);
 	private Pattern maxAgePattern = Pattern.compile("max-age=(\\d*)");
-	private ReentrantLock keysLock = new ReentrantLock();
 	private volatile Long expires = 0L;
 	private Map<String, JwtDecoder> delegates = new ConcurrentHashMap<>();
+
 	public FirebaseJwtTokenDecoder(RestOperations restClient, String googlePublicKeysEndpoint, OAuth2TokenValidator<Jwt> tokenValidator) {
 		this.restClient = restClient;
 		this.googlePublicKeysEndpoint = googlePublicKeysEndpoint;
 		this.tokenValidator = tokenValidator;
 	}
+
 	@Override
-	public Jwt decode(String token) throws JwtException {
+	public synchronized Jwt decode(String token) throws JwtException {
```

Comment:
We can decrease the synchronized surface by moving `synchronized` to the `refresh()` method -- that's what was attempted to be protected by the original lock.

The decoding itself can be done in parallel.

Reply:
Good call - done.